MiniApp/Classes/MiniAppDownloader.swift

```diff
@@ -22,15 +22,26 @@ class MiniAppDownloader {
     }
 
     func download(appId: String, versionId: String, completionHandler: @escaping (Result<URL, Error>) -> Void) {
-        let miniAppStoragePath = FileManager.getMiniAppDirectory(with: appId, and: versionId)
-        if miniAppStatus.isDownloaded(key: "\(appId)/\(versionId)") {
+        let miniAppStoragePath = FileManager.getMiniAppVersionDirectory(with: appId, and: versionId)
+        if miniAppStatus.isDownloaded(appId: appId, versionId: versionId) {
             completionHandler(.success(miniAppStoragePath))
             return
         }
         self.manifestDownloader.fetchManifest(apiClient: self.miniAppClient, appId: appId, versionId: versionId) { (result) in
             switch result {
             case .success(let responseData):
-                self.downloadMiniApp(urls: responseData.manifest, to: miniAppStoragePath, completionHandler: completionHandler)
+                self.downloadMiniApp(urls: responseData.manifest, to: miniAppStoragePath) { downloadResult in
+                    switch downloadResult {
+                    case .success:
+                        if let error = self.miniAppStorage.cleanVersions(for: appId, differentFrom: versionId, status: self.miniAppStatus) {
```

Comment:
Wouldn't this add delay to provide the display unit back to consumer app? 
It would be a good idea to decouple it.

Reply:
yes if they are working on a potato 🥔 . But in this case, wouldn't it be harder to them to display a miniApp while older versions are deleted (system lagging) ? It's not sarcasm, it's a real question

Reply:
On a potato? I missed some context. :)

I am not 100% sure how iPhones respond to parallel execution but I do think they have the cores to do it. 😉  A mini app is allowed to be 50MB in size, can easily have thousands of files (like in panda park, disk I/O is expensive).

Reply:
I think it would be a great idea to move them into [Operation](https://developer.apple.com/documentation/foundation/operation) or Dispatch block and immediately return the display unit. Because few mini-apps files/structure would be complex (it could even cause one-second delay) to delete it.

Reply:
it's not really a threading issue, it's more about memory throttling on old devices

Reply:
it's not really a threading issue, it's more about memory throttling on old devices

Reply:
By the way, it doesn't hurt to put it into an async thread ;)

Reply:
The display doesn't need to be blocked by deletion.

Reply:
> On a potato? I missed some context. :)
> 
> I am not 100% sure how iPhones respond to **parallel** execution but I do think they have the _cores_ to do it. 😉 A mini app is allowed to be 50MB in size, can easily have thousands of files (like in panda park, disk I/O is expensive).

A potato is a very old device on end of life, just a joke :) Well on very old devices, with cheap/end of life memory, accessing storage memory can be really problematic, and it can be problematic to launch new version of PandaPark while its old version is being deleted. I'm not talking about iPhone X, and you're both right it's better to thread this, but user experience on old device can be degraded far more than waiting a few more seconds.